src/nupic/algorithms/ColumnPooler.cpp

```diff
@@ -0,0 +1,565 @@
+    for(auto i = 0u; i < tieBreaker_.size(); ++i) {
+      tieBreaker_[i] = 0.01f * rng_.getReal64();
+    }
+    proximalMaxSegment_.resize( cells.size );
+    AF = new SDR_ActivationFrequency( {cells.size, proximalSegments}, period );
+    AF->initializeToValue( sparsity / proximalSegments );
+
+    // Setup the distal dendrites
+    distalConnections.initialize(
+        /* columnDimensions */            {2048},
+        /* cellsPerColumn */              1,
+        /* activationThreshold */         distalSegmentThreshold,
+        /* initialPermanence */           0.21, // TODO: Needs to be a parameter!!!
+        /* connectedPermanence */         distalSynapseThreshold,
+        /* minThreshold */                distalSegmentMatch,
+        /* maxNewSynapseCount */          distalAddSynapses,
+        /* permanenceIncrement */         distalIncrement,
+        /* permanenceDecrement */         distalDecrement,
+        /* predictedSegmentDecrement */   distalMispredictDecrement,
+        /* seed */                        rng_(),
+        /* maxSegmentsPerCell */          distalMaxSegments,
+        /* maxSynapsesPerSegment */       distalMaxSynapsesPerSegment,
+        /* checkInputs */                 true,
+        /* extra */                       0);
+
+    iterationNum_      = 0u;
+    iterationLearnNum_ = 0u;
+
+    reset();
+
+    if( PP_Sp.min() * proximalInputs.size < proximalSegmentThreshold )
+      cerr << "WARNING: Proximal segment has fewer synapses than the segment threshold." << endl;
+    if( PP_AF.min() == 0.0f )
+      cerr << "WARNING: Proximal input is unused." << endl;
+
+    if( verbose ) {
+      // TODO: Print all parameters
+      cout << "Potential Pool Statistics:" << endl
+           << PP_Sp
+           << PP_AF << endl;
+    }
+  }
+
+
+  void reset() {
+    X_act.assign( proximalConnections.numCells(), 0.0f );
+    X_inact.assign( proximalConnections.numCells(), 0.0f );
+    // TODO Zero Previous Updates
+    distalConnections.reset();
+  }
+
+
+  void compute(
+        const SDR& proximalInputActive,
+        bool learn,
+        SDR& active) {
+    SDR none( distalInputDimensions );
+    SDR none2( active.dimensions );
+    compute(proximalInputActive, proximalInputActive, none, none, learn, active, none2 );
+  }
+
+  void compute(
+        const SDR& proximalInputActive,
+        const SDR& proximalInputLearning,
+        const SDR& distalInputActive,
+        const SDR& distalInputLearning,
+        bool learn,
+        SDR& active,
+        SDR& learning) {
+    NTA_CHECK( proximalInputActive.dimensions   == proximalInputDimensions );
+    NTA_CHECK( proximalInputLearning.dimensions == proximalInputDimensions );
+    NTA_CHECK( distalInputActive.dimensions     == distalInputDimensions );
+    NTA_CHECK( distalInputLearning.dimensions   == distalInputDimensions );
+    NTA_CHECK( active.dimensions                == cellDimensions );
+    NTA_CHECK( learning.dimensions              == cellDimensions );
```

Comment:
consider if these could be just NTA_ASSERTs

Reply:
This is the main user-facing API method.  Detecting bad arguments is a feature.  When we ship this, it will probably be in release mode.  

These checks should be fast.  They compare vectors but the vectors are typically only a few numbers long.  These checks are at the outer most loop so they happen once per call.